### ISSUE TITLE
some questions about PredictionEngine and others

### ISSUE THREAD

**Issue by magicdict** [dotnet/machinelearning#3890]
### System information

- Win7:
- NetCore 3.0 Preview6

### Issue

- **What did you do?**
Try to create PredictionEngine

- **What happened?**

An unhandled exception of type 'System.ArgumentOutOfRangeException' occurred in Microsoft.ML.Core.dll: 'Features column 'Feature' not found'

```CSharp
PredictionEngine<Survey, SurveyPrediction> predictionFunction =
             mlContext.Model.CreatePredictionEngine<Survey, SurveyPrediction>

```
- **What did you expect?**

predictionFunction is create 

### Source code / logs

```Csharp
using System;
using System.Collections.Generic;
using Microsoft.ML;
using Microsoft.ML.Data;
using Microsoft.ML.Trainers.LightGbm;

namespace Happiness
{
    class Program
    {
        static void Main(string[] args)
        {
            var mlContext = new MLContext(seed: 2019);
            var trainingDataView = mlContext.Data.LoadFromTextFile<Survey>(@"F:\Happiness\data\happiness_train_abbr.csv", hasHeader: true, separatorChar: ',');
            var SubmitDataView = mlContext.Data.LoadFromTextFile<SurveyTest>(@"F:\Happiness\data\happiness_test_abbr.csv", hasHeader: true, separatorChar: ',');

            //数据准备
            //特征项目
            var dataProcessPipeline = mlContext.Transforms.Concatenate("Features", nameof(Survey.birth),
                                                                                   nameof(Survey.car),
                                                                                   nameof(Survey.class_1),
                                                                                   nameof(Survey.county)
                                                                                   );

            var DataTransformer = dataProcessPipeline.Fit(trainingDataView);
            trainingDataView = DataTransformer.Transform(trainingDataView);
            SubmitDataView = DataTransformer.Transform(SubmitDataView);

            DataOperationsCatalog.TrainTestData dataSplit = mlContext.Data.TrainTestSplit(trainingDataView, testFraction: 0.2);
            IDataView trainData = dataSplit.TrainSet;
            IDataView testData = dataSplit.TestSet;

            IEnumerable<Survey> HappinessDataEnumerable =
                mlContext.Data.CreateEnumerable<Survey>(trainData, reuseRowObject: true);

            int MaxCnt = 10;
            int Current = 0;
            // Iterate over each row
            foreach (Survey row in HappinessDataEnumerable)
            {
                // Do something (print out Size property) with current Housing Data object being evaluated
                Console.WriteLine(row.id + ":" + row.height_cm + "," + row.happiness);
                Current++;
                if (Current == MaxCnt) break;
            }


            // Define trainer options.
            var options = new LightGbmRegressionTrainer.Options
            {
                LabelColumnName = nameof(Survey.happiness),
                FeatureColumnName = "Features",
                // How many leaves a single tree should have.
                NumberOfLeaves = 64,
                NumberOfIterations = 100,
                EvaluationMetric = LightGbmRegressionTrainer.Options.EvaluateMetricType.RootMeanSquaredError,
                LearningRate = 0.001,
                Booster = new Microsoft.ML.Trainers.LightGbm.GossBooster.Options()
                {
                    TopRate = 0.3,
                    OtherRate = 0.2
                }
            };

            var LightGBMEstimator = mlContext.Regression.Trainers.LightGbm(options);
            var LightGBMTransformer = LightGBMEstimator.Fit(trainData);
            testData = LightGBMTransformer.Transform(testData);
            var metrics = mlContext.Regression.Evaluate(testData, labelColumnName: nameof(Survey.happiness));
            Common.ConsoleHelper.PrintRegressionMetrics(LightGBMEstimator.ToString(), metrics);

            PredictionEngine<Survey, SurveyPrediction> predictionFunction =
             mlContext.Model.CreatePredictionEngine<Survey, SurveyPrediction>(LightGBMTransformer);
            var Submit = LightGBMTransformer.Transform(SubmitDataView);
            HappinessDataEnumerable =
                mlContext.Data.CreateEnumerable<Survey>(Submit, reuseRowObject: true);

            Current = 0;
            foreach (Survey row in HappinessDataEnumerable)
            {
                // Do something (print out Size property) with current Housing Data object being evaluated
                var predictions = predictionFunction.Predict(row);
                Console.WriteLine(predictions.id + ":" + predictions.height_cm + "," + predictions.Prediction);
                Current++;
                if (Current == MaxCnt) break;
            }


        }
    }
}

```

The LightGBM look can work and I get the Metrics,so I think the model is no problem...
```shell
*************************************************
*       Metrics for Microsoft.ML.Trainers.LightGbm.LightGbmRegressionTrainer regression model      
*------------------------------------------------
*       LossFn:        1.09
*       R2 Score:      0.02
*       Absolute loss: .61
*       Squared loss:  1.09
*       RMS loss:      1.04
*************************************************
```

**Comment by magicdict** [comment]
Q2:How to set Chinese RemoveDefaultStopWords?
I Guess I can load a stop word list form a external file.

Q3: Sometimes,the train dataset has the same layout with the test dataset,but the test dataset has no label column, I want to load from csv file,I have to define two class only the loadcolumn is diffrent.
Is there a good solution or load csv file by tradtional streamreader....

happiness is Label Column,traning dataset has the column,but test dataset has not. 

```CSharp
public class SurveyRawData
{
    public const int FirstIndex = 2;

    [LoadColumn(0)]
    public string id;
    [LoadColumn(1)]
    public float happiness;
    [LoadColumn(FirstIndex)]
    public float survey_type;
    [LoadColumn(FirstIndex + 1)]
    public float province;
    [LoadColumn(FirstIndex + 2)]
}

public class SurveyRawDataTest
{
    public const int FirstIndex = 1;
    [LoadColumn(41)]
    public float happiness;
    [LoadColumn(0)]
    public string id;
    [LoadColumn(FirstIndex)]
    public float survey_type;
    [LoadColumn(FirstIndex + 1)]
    public float province;
    [LoadColumn(FirstIndex + 2)]
}
```

**Comment by magicdict** [comment]
Q4.
Why trainer has  Fit(train, validationDv) Method ,but trainingPipeline has not?

```CSharp
var trainer = mlContext.Regression.Trainers.LightGbm(options);
var trainingPipeline = dataProcessPipeline.Append(trainer);
var trainedModel = trainingPipeline.Fit(folds[i].TrainSet);    //No Fit(train, validationDv) Method 
```

**Comment by magicdict** [comment]
Q5.
```CSharp
 var LightGBMEstimator = mlContext.Regression.Trainers.LightGbm(options);
            var LightGBMTransformer = LightGBMEstimator.Fit(trainData);
            var AftertestData = LightGBMTransformer.Transform(testData);
```

AftertestData has a new Column ’Score‘ ，  is it predict value?

What is a 'PredictedLabel’ ? used by PredictionEngine?

**Comment by najeeb-kazmi** [comment]
@magicdict 

Q1: `LightGBMTransformer` is fit on the `trainData` after all the transformations have been applied. At this point, it expects a `Features` column. In `mlContext.Model.CreatePredictionEngine<Survey, SurveyPrediction>`, the `Survey` class has no `Features` column. I would change your code to use one pipeline of multiple estimators:
```csharp
var mlContext = new MLContext(seed: 2019);
var trainingDataView = mlContext.Data.LoadFromTextFile<Survey>(@"F:\Happiness\data\happiness_train_abbr.csv", hasHeader: true, separatorChar: ',');
var SubmitDataView = mlContext.Data.LoadFromTextFile<SurveyTest>(@"F:\Happiness\data\happiness_test_abbr.csv", hasHeader: true, separatorChar: ',');

// Split first
DataOperationsCatalog.TrainTestData dataSplit = mlContext.Data.TrainTestSplit(trainingDataView, testFraction: 0.2);
IDataView trainData = dataSplit.TrainSet;
IDataView testData = dataSplit.TestSet;

// Then build pipeline
var pipeline =
    mlContext.Transforms.Concatenate("Features", nameof(Survey.birth),
                                                                               nameof(Survey.car),
                                                                               nameof(Survey.class_1),
                                                                               nameof(Survey.county)
                                                                               )
    .Append(mlContext.Regression.Trainers.LightGbm(options));

var transformer = pipeline.Fit(trainData);
PredictionEngine<Survey, SurveyPrediction> predictionFunction =
             mlContext.Model.CreatePredictionEngine<Survey, SurveyPrediction>(LightGBMTransformer);
```

**Comment by najeeb-kazmi** [comment]
Q2: You can remove custom stopwords by specifying a custom array. See [this sample](https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/Text/RemoveStopWords.cs).
https://github.com/dotnet/machinelearning/blob/3b69b0da07a8298abda6559a81fa48c9a569351d/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/Text/RemoveStopWords.cs#L29-L33

**Comment by najeeb-kazmi** [comment]
@magicdict 

Q3: If the label is the last column of your file, you can read the data in with just one class. For example, if you have [this version of the Iris dataset](https://github.com/dotnet/machinelearning/blob/master/test/data/iris.data), where the label is the last column, you can simply use the following class to load both training data with label and test data without label.
```csharp
        var trainData = mlContext.Data.LoadFromTextFile<IrisInput>(trainPath);
        // test data does not have label colum, only four feature columns
        var testData = mlContext.Data.LoadFromTextFile<IrisInput>(testPath);

        private class IrisInput
        {
            [LoadColumn(0)]
            public float SepalLength { get; set; }

            [LoadColumn(1)]
            public float SepalWidth { get; set; }

            [LoadColumn(2)]
            public float PetalLength { get; set; }

            [LoadColumn(3)]
            public float PetalWidth { get; set; }

            [LoadColumn(4)]
            public string Label { get; set; }
        }
```

If the label column is not the last column, but is the first column (or any other column in the middle), you can read the data using the same class if you ensure that there is an empty column in the file. For example, if you have the following training data instead where the delimiter is `,`

`Iris-setosa,5.1,3.5,1.4,0.2`

and the test data doesn't have the label, ensure that the first column is indicated to be empty by use of the delimiter:
`,5.1,3.5,1.4,0.2`

then you can use the following class to load both data:

```csharp
        private class IrisInput
        {
            [LoadColumn(0)]
            public string Label { get; set; }

            [LoadColumn(1)]
            public float SepalLength { get; set; }

            [LoadColumn(2)]
            public float SepalWidth { get; set; }

            [LoadColumn(3)]
            public float PetalLength { get; set; }

            [LoadColumn(4)]
            public float PetalWidth { get; set; }
        }
```

**Comment by najeeb-kazmi** [comment]
@magicdict 

Q4:
`LightGbm` is one of the few trainers that supports validation set. When composing a pipeline with multiple components, you create a `TransformerChain` object. This object does not have a `.Fit` method that accepts both training and validation set as it has to generalize to the base estimators that do not have validation dataset support.

**Comment by najeeb-kazmi** [comment]
@magicdict 

Q5:
Yes, `Score` column is the predicted value for regression trainers. The 'PredictedLabel' used by `PredictionEngine` would be an output class with a `Score` property:
```csharp
private class RegresionPrediction
{
        public float Score { get; set; }
}
```

**Comment by najeeb-kazmi** [comment]
@magicdict I'm closing this issue as all the questions have been answered. Please feel free to reopen if you have more questions.
